### PR TITLE
Font Library: create font faces through the REST API

### PR DESCRIPTION
**Pull request by creativecoder**
## What?

Continuing the work started in https://github.com/WordPress/gutenberg/pull/57656 to build out the Font Faces REST API controller

- Returning proper response data
- Creating font faces, including uploading files

## Testing Instructions

Test the following routes and ensure the proper responses

```
GET wp/v2/font-families/<id>/font-faces
GET wp/v2/font-families/<id>/font-faces/<id>
POST wp/v2/font-families/<id>/font-faces
DELETE wp/v2/font-families/<id>/font-faces/<id>
```

The POST request requires using multipart/form-data to include a font file. Here's an example of using the Postman client to make a request

image

(you can "Copy as curl" from your browser network tab to get the cookies, and then "Import" that into Postman to make an authenticated request)

### CODE REVIEW

File: lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php [hunk 1]
```diff
@@ -228,7 +228,99 @@ public function get_item( $request ) {
...
+	protected function handle_font_file_upload( $file ) {
+		add_filter( 'upload_mimes', array( 'WP_Font_Library', 'set_allowed_mime_types' ) );
+		add_filter( 'upload_dir', array( 'WP_Font_Library', 'set_upload_dir' ) );
+
+		$overrides = array(
+			// Not testing a form submission.
+			'test_form' => false,
+			// Seems mime type for files that are not images cannot be tested.
+			// See wp_check_filetype_and_ext().
+			'test_type' => false,
```
Comment (pbking) [inline lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php]: The mime times passed via this overrides object don't seem to matter, however the `set_allowed_mime_types` filter seems to achieve what we need allowing the mime types to be tested.

I believe the above value should be true, however I think the 'mimes' value can be eliminated.

File: lib/experimental/fonts/font-library/font-library.php [hunk 2]
```diff
@@ -118,3 +118,26 @@ function wp_unregister_font_collection( $collection_id ) {
...
+ *
+ * The path is relative to the current fonts dir.
+ *
+ * @since 6.5.0
+ * @access private
+ *
+ * @param string $path Full path to the file.
+ * @return string Relative path on success, unchanged path on failure.
+ */
+function _wp_relative_fonts_path( $path ) {
```
Comment (pbking) [inline lib/experimental/fonts/font-library/font-library.php]: Until this has been shown to be useful elsewhere perhaps better to make this a private function of the class instead?

File: lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php [hunk 3]
```diff
@@ -228,7 +228,99 @@ public function get_item( $request ) {
...
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function create_item( $request ) {
+		// Create the font face post so we have an ID to attach the font files to.
+		$font_face_id = wp_insert_post(
+			array(
+				'post_type'   => $this->post_type,
+				'post_parent' => $request['parent'],
+				'post_status' => 'publish',
+				'post_title'  => 'Font Face',
```
Comment (matiasbenedetto) [inline lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php]: Maybe we could use something more meaningful here as the `fontFamily` value?

File: lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php [hunk 4]
```diff
@@ -255,6 +347,30 @@ public function delete_item( $request ) {
...
+	 *
+	 * @param WP_Post         $item    Post object.
+	 * @param WP_REST_Request $request Request object.
+	 * @return WP_REST_Response Response object.
+	 */
+	public function prepare_item_for_response( $item, $request ) {
+		$data = array();
+
+		$data['id']                 = $item->ID;
+		$data['theme_json_version'] = 2;
```
Comment (matiasbenedetto) [inline lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:337]:
```suggestion
		$data['theme_json_version'] = WP_Theme_JSON::LATEST_SCHEMA;
```

File: phpunit/tests/fonts/font-library/wpRestFontFacesController.php [hunk 5]
```diff
@@ -309,4 +388,25 @@ public function test_get_item_schema() {
...
+		$post = get_post( $post_id );
+
+		$this->assertArrayHasKey( 'id', $data );
+		$this->assertSame( $post->ID, $data['id'] );
+
+		$this->assertArrayHasKey( 'parent', $data );
+		$this->assertSame( $post->post_parent, $data['parent'] );
+
+		$this->assertArrayHasKey( 'theme_json_version', $data );
+		$this->assertSame( 2, $data['theme_json_version'] );
```
Comment (matiasbenedetto) [inline phpunit/tests/fonts/font-library/wpRestFontFacesController.php]:
```suggestion
		$this->assertSame( WP_Theme_JSON::LATEST_SCHEMA, $data['theme_json_version'] );
```

**Review by matiasbenedetto — COMMENTED** [review]
I think the tasks regarding creating the font faces posts and assets should not be handled in the rest controller but in another class. It would make sense for that other class to be named `WP_Font_Face`, but a class with that name already exists, and it is in charge of printing the Font Face CSS. Personally, I think that the work needed to merge the 2 classes doesn't make sense right now. Should pick a new name like `WP_Font_Library_Face` ?

File: lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php (hunk 3 above)
Comment (creativecoder) [reply lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php]: Idea: use post_name for font family

Note that font family currently uses font name for post_title, and font slug for post_name.

File: lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php (hunk 4 above)
Comment (creativecoder) [reply lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:337]: Idea: tests should break if latest theme.json version is changed.

File: lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php (hunk 1 above)
Comment (creativecoder) [reply lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php]: Thanks! It does seem to work without the `mimes` override here, so I've removed it.

File: lib/experimental/fonts/font-library/font-library.php (hunk 2 above)
Comment (creativecoder) [reply lib/experimental/fonts/font-library/font-library.php]: Done.

File: lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php (hunk 3 above)
Comment (creativecoder) [reply lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php]: Done, now using `fontFamily` for post_title, and a generated slug from that for post_name.

File: lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php (hunk 4 above)
Comment (creativecoder) [reply lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:337]: I've changed check_font_face_data to use  `WP_Theme_JSON::LATEST_SCHEMA` when checking the response data, so tests will fail if the theme.json version increases.

**Comment by creativecoder** [comment]
> I think the tasks regarding creating the font faces posts and assets should not be handled in the rest controller but in another class. 

I'm not sure that's needed or adds much benefit with the current implementation, but I'm open to it. I'd like to get the font faces and font families controller working with the new API design, using patterns from other Core controllers as much as possible, and then do a pass to refactor and optimize.

**Comment by matiasbenedetto** [comment]
> I'm not sure that's needed or adds much benefit with the current implementation

It adds the ability to create a PHP api to create the font faces. Something like `wp_install_font_face(...)` or `WP_Font_Face->install()` calling a rest controller does not seem like the most logical approach.

**Review by mikachan — APPROVED** [review]
Great job, this is testing really well for me!

Test report with happy paths:

- `GET wp/v2/font-families/<id>/font-faces` ✅
- `GET wp/v2/font-families/<id>/font-faces/<id>` ✅
- `POST wp/v2/font-families/<id>/font-faces` ✅
- `DELETE wp/v2/font-families/<id>/font-faces/<id>` ✅

Error testing:

`validate_create_font_face_settings`:
- I saw an error when I tried sending no font file ✅
- I saw an error with sending an empty `src` ✅

`GET` requests:
- I see errors if the parent font family doesn't exist, or the font face ID doesn't exist ✅
- I see an error if the font face does not belong to the specified font family id ✅

**Pull request merged**